src/core/utils/compatibility.js

```diff
@@ -101,6 +102,11 @@ export function isCompatibleWithUserAgent({
 
     if (
       addon.type === ADDON_TYPE_OPENSEARCH &&
+      // If window isn't defined but the browser is Firefox, this is likely a
+      // server-side render, so we shouldn't condemn this search plugin to
+      // incompatibility yet, or it will stay disabled.
+      // See: https://github.com/mozilla/addons-frontend/issues/4047
+      _hasWindow &&
       !(_window.external && 'AddSearchProvider' in _window.external)
```

Comment:
I think you've uncovered a deeper bug here. It looks like React is rendering server HTML for the button component but not re-rendering it on the client (as it should). I forget exactly how this works but I think it has to do with the way React diffs prop values.

After your patch, the bug isn't fixed. Or, rather, there's a new bug where the button appears enabled for any Firefox client that does *not* define `window.external.AddSearchProvider()`. It should appear disabled.

I was testing it with a patch to simulate a missing `AddSearchProvider` API:

```diff
diff --git a/src/core/utils/compatibility.js b/src/core/utils/compatibility.js
index edfc29d5e..b23722140 100644
--- a/src/core/utils/compatibility.js
+++ b/src/core/utils/compatibility.js
@@ -107,7 +107,7 @@ export function isCompatibleWithUserAgent({
       // incompatibility yet, or it will stay disabled.
       // See: https://github.com/mozilla/addons-frontend/issues/4047
       _hasWindow &&
-      !(_window.external && 'AddSearchProvider' in _window.external)
+      !(_window.external && 'AddSearchProviderNOPE' in _window.external)
     ) {
       return { compatible: false, reason: INCOMPATIBLE_NO_OPENSEARCH };
     }

```

<img width="608" alt="screenshot 2018-02-01 14 22 25" src="https://user-images.githubusercontent.com/55398/35701488-cb5afbbe-075b-11e8-9940-e30d3ac23406.png">

I have a feeling this may be hard to fix 😄 If that's the case, I would suggest filing a bug to fix it later because it's not so important to break Firefox clients without `AddSearchProvider` -- do they even exist?

Reply:
I think the only client that would identify itself as Firefox but not have the `AddSearchProvider` API is a REALLY old version (unsure) or Firefox for iOS. The latter we care more about and would be covered by a broader incompatibility check so it should be fine.